### PR TITLE
Add active clients -trainers/validators with roles

### PR DESCRIPTION
**Pull request by aitmlouk**
## Status

- [x] Ready
- [ ] Draft
- [ ] Hold

## Description

Describe your changes here to communicate to the maintainers why we should accept this pull request. 
If it fixes a bug or resolves a feature request, be sure to include a link to that issue.

## Types of changes

What types of changes does your code introduce to FEDn?

- [ ] Hotfix (fixing a critical bug in production)
- [ ] Bugfix
- [x] New feature
- [ ] Documentation update

## Checklist

_If you're unsure about any of the items below, don't hesitate to ask. We're here to help! 
This is simply a reminder of what we are going to look for before merging your code._

- [ ] This pull request is against **develop** branch (not applicable for hotfixes)
- [ ] I have included a link to the issue on GitHub or JIRA (if any)
- [ ] I have included migration files (if there are changes to the model classes)
- [ ] I have read the [CONTRIBUTING](https://github.com/scaleoutsystems/fedn/blob/master/CONTRIBUTING.md) doc
- [ ] I have included tests to complement my changes
- [ ] I have updated the related documentation (if necessary) 
- [ ] I have added a reviewer for this pull request
- [ ] I have added myself as an author for this pull request

## Further comments

Anything else you think we should know before merging your code!

### CODE REVIEW

File: fedn/fedn/clients/reducer/restservice.py [hunk 1]
```diff
@@ -477,8 +477,45 @@ def combiner_stats():
                     combiner_info.append(report)
                 except:
                     pass
-                return combiner_info
-            return False
+            return combiner_info
+
+        def client_stats():
```
Comment (ahellander) [inline fedn/fedn/clients/reducer/restservice.py:482]: Minor detail, but can we call it "client_status" ?

File: fedn/fedn/clients/reducer/restservice.py [hunk 2]
```diff
@@ -477,8 +477,45 @@ def combiner_stats():
...
+            client_info = self.control.get_client_info()
+            active_clients = combiner_stats()
+
+            for client in active_clients:
+                active_trainers_str = client['active_trainers']
+                active_validators_str = client['active_validators']
+
+            import re
```
Comment (ahellander) [inline fedn/fedn/clients/reducer/restservice.py:490]: Move import to top of file, module level.

File: fedn/fedn/clients/reducer/restservice.py [hunk 3]
```diff
@@ -477,8 +477,45 @@ def combiner_stats():
...
+            active_trainers_list = ' '.join(active_trainers_str.split(" ")).split()
+
+            active_validators_str = re.sub('[^a-zA-Z0-9:\n\.]', '', active_validators_str).replace('name:', ' ')
+            active_validators_list = ' '.join(active_validators_str.split(" ")).split()
+
+            active_trainers_list_ = [client for client in client_info if client['name'] in active_trainers_list]
+            active_validators_list_ = [cl for cl in client_info if cl['name'] in active_validators_list]
+
+            if active_trainers_list_ == active_validators_list_:
+                print('ACTIVE CLIENTS ARE ALSO VALIDATORS')
```
Comment (ahellander) [inline fedn/fedn/clients/reducer/restservice.py:501]: Remove print statements.

**Review by ahellander — CHANGES_REQUESTED** [review]
Great. Added some small edits.

File: fedn/fedn/clients/reducer/control.py [hunk 4]
```diff
@@ -74,12 +74,18 @@ def get_latest_model(self):
     def get_model_info(self):
         return self.statestore.get_model_info()
 
+    def get_client_info(self):
```
Comment (ahellander) [inline fedn/fedn/clients/reducer/control.py:77]: I think this should belong in network.py

File: fedn/fedn/clients/reducer/control.py [hunk 5]
```diff
@@ -74,12 +74,18 @@ def get_latest_model(self):
...
+        return self.statestore.list_clients()
+
     def get_events(self):
         return self.statestore.get_events()
 
     def drop_models(self):
         self.statestore.drop_models()
 
+    def update_client_data(self, client_data, status, role):
```
Comment (ahellander) [inline fedn/fedn/clients/reducer/control.py:86]: I think this should belong in network.py

File: fedn/fedn/clients/reducer/restservice.py [hunk 6]
```diff
@@ -479,7 +480,65 @@ def combiner_stats():
                 except:
                     pass
             return combiner_info
-            return False
+
+        def client_status():
```
Comment (ahellander) [inline fedn/fedn/clients/reducer/restservice.py:484]: docstring missing

**Review by ahellander — CHANGES_REQUESTED** [review]

File: fedn/fedn/clients/reducer/control.py (hunk 4 above)
Comment (aitmlouk) [reply fedn/fedn/clients/reducer/control.py]: I see, done!

File: fedn/fedn/clients/reducer/control.py (hunk 5 above)
Comment (aitmlouk) [reply fedn/fedn/clients/reducer/control.py]: done!

File: fedn/fedn/clients/reducer/restservice.py (hunk 6 above)
Comment (aitmlouk) [reply fedn/fedn/clients/reducer/restservice.py:484]: hmm, fixed maybe we need also to check docstring for the whole framework

Comment (ahellander) [reply fedn/fedn/clients/reducer/restservice.py:484]: Yep. Prashant has a PR that adds docstrings everywhere.

**Review by ahellander — APPROVED** [review]

**Pull request merged**